src/main/java/org/prebid/server/bidder/nanointeractive/NanointeractiveBidder.java

```diff
@@ -0,0 +1,162 @@
+package org.prebid.server.bidder.nanointeractive;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Device;
+import com.iab.openrtb.request.Imp;
+import com.iab.openrtb.request.Site;
+import com.iab.openrtb.response.BidResponse;
+import com.iab.openrtb.response.SeatBid;
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpMethod;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpCall;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.nanointeractive.ExtImpNanointeractive;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class NanointeractiveBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpNanointeractive>> NANOINTERACTIVE_EXT_TYPE_REFERENCE =
+            new TypeReference<ExtPrebid<?, ExtImpNanointeractive>>() {
+            };
+    private static final String DEFAULT_BID_CURRENCY = "USD";
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public NanointeractiveBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<Imp> validImps = new ArrayList<>();
+        String reference = "";
+        for (Imp imp : request.getImp()) {
+            try {
+                final Imp validImp = validateImp(imp);
+                final ExtImpNanointeractive extImp = parseImpExt(imp);
+                if (StringUtils.isBlank(extImp.getPid())) {
+                    throw new PreBidException("pid is empty");
+                }
+
+                if (StringUtils.isNotBlank(extImp.getRef())) {
```

Comment:
You will use last not empty `extImp.getRef()`. But in GO they will use first one.